File: ios/HybridFacebookAnalytics.swift
import Foundation
import NitroModules
import FBSDKCoreKit
import AdSupport

class HybridFacebookAnalytics: HybridFacebookAnalyticsSpec {

    // MARK: - Initialization & Runtime Configuration

    func initialize() throws {
        // On iOS, auto-init is controlled by Info.plist (FacebookAutoInitEnabled).
        // When disabled, this performs the deferred initialization.
        ApplicationDelegate.shared.initializeSDK()
    }

    func setAutoLogAppEventsEnabled(enabled: Bool) throws {
        Settings.shared.isAutoLogAppEventsEnabled = enabled
    }

    func setAdvertiserIDCollectionEnabled(enabled: Bool) throws {
        Settings.shared.isAdvertiserIDCollectionEnabled = enabled
    }

    func setAppID(appID: String) throws {
        Settings.shared.appID = appID
    }

    func setClientToken(clientToken: String) throws {
        Settings.shared.clientToken = clientToken
    }

    func setLoggingEnabled(enabled: Bool) throws {
        let behaviors: [LoggingBehavior] = [
            .appEvents, .networkRequests, .developerErrors, .informational,
        ]
        for behavior in behaviors {
            if enabled {
                Settings.shared.enableLoggingBehavior(behavior)
            } else {
                Settings.shared.disableLoggingBehavior(behavior)
            }
        }
    }

    // MARK: - Event Logging

    func logEvent(eventName: String, valueToSum: Double, params: [String: String]) throws {
        let fbParams: [AppEvents.ParameterName: Any] = params.reduce(into: [:]) { result, pair in
            result[AppEvents.ParameterName(pair.key)] = pair.value
        }
        AppEvents.shared.logEvent(
            AppEvents.Name(eventName),
            valueToSum: valueToSum,
            parameters: fbParams
        )
    }

    func logEventWithoutParams(eventName: String, valueToSum: Double) throws {
        if valueToSum != 0 {
            AppEvents.shared.logEvent(
                AppEvents.Name(eventName),
                valueToSum: valueToSum
            )
        } else {
            AppEvents.shared.logEvent(AppEvents.Name(eventName))
        }
    }

    func logPurchase(amount: Double, currency: String, params: [String: String]) throws {
        let fbParams: [AppEvents.ParameterName: Any] = params.reduce(into: [:]) { result, pair in
            result[AppEvents.ParameterName(pair.key)] = pair.value
        }
        AppEvents.shared.logPurchase(amount: amount, currency: currency, parameters: fbParams)
    }

    func logPushNotificationOpen(payload: [String: String]) throws {
        let nsPayload: [String: Any] = payload.reduce(into: [:]) { result, pair in
            result[pair.key] = pair.value
        }
        AppEvents.shared.logPushNotificationOpen(payload: nsPayload)
    }

    // MARK: - User Identity

    func setUserID(userID: String) throws {
        AppEvents.shared.userID = userID
    }

    func clearUserID() throws {
        AppEvents.shared.userID = nil
    }

    func getUserID() throws -> String? {
        return AppEvents.shared.userID
    }

    func setUserData(
        email: String?,
        firstName: String?,
        lastName: String?,
        phone: String?,
        dateOfBirth: String?,
        gender: String?,
        city: String?,
        state: String?,
        zip: String?,
        country: String?
    ) throws {
        AppEvents.shared.setUser(
            email: email,
            firstName: firstName,
            lastName: lastName,
            phone: phone,
            dateOfBirth: dateOfBirth,
            gender: gender,
            city: city,
            state: state,
            zip: zip,
            country: country
        )
    }

    // MARK: - Device Identifiers

    func getAnonymousID() throws -> Promise<String?> {
        return Promise.resolved(withResult: AppEvents.shared.anonymousID)
    }

    func getAdvertiserID() throws -> Promise<String?> {
        let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
        if idfa == "00000000-0000-0000-0000-000000000000" {
            return Promise.resolved(withResult: nil)
        }
        return Promise.resolved(withResult: idfa)
    }

    // MARK: - Configuration

    func setFlushBehavior(flushBehavior: String) throws {
        switch flushBehavior {
        case "explicit_only":
            AppEvents.shared.flushBehavior = .explicitOnly
        default:
            AppEvents.shared.flushBehavior = .auto
        }
    }

    func flush() throws {
        AppEvents.shared.flush()
    }

    // MARK: - Push Notifications

    func setPushNotificationsDeviceToken(token: String) throws {
        // Token comes as a hex string — convert back to raw Data.
        // Iterate over UTF-8 bytes and pack each pair of hex digits into one byte.
        var data = Data(capacity: token.utf8.count / 2)
        var highNibble: UInt8? = nil
        for ascii in token.utf8 {
            let nibble: UInt8
            switch ascii {
            case 0x30...0x39: nibble = ascii - 0x30        // '0'-'9'
            case 0x41...0x46: nibble = ascii - 0x41 + 10   // 'A'-'F'
            case 0x61...0x66: nibble = ascii - 0x61 + 10   // 'a'-'f'
            default: continue                              // skip spaces / non-hex
            }
            if let high = highNibble {
                data.append((high << 4) | nibble)
                highNibble = nil
            } else {
                highNibble = nibble
            }
        }
        AppEvents.shared.setPushNotificationsDeviceToken(data)
    }

    func setPushNotificationsRegistrationId(registrationId: String) throws {
        // No-op on iOS — this is an Android-only method
    }
}

File: android/src/main/java/com/nitrofbanalytics/HybridFacebookAnalytics.kt
package com.nitrofbanalytics

import android.os.Bundle
import com.facebook.FacebookSdk
import com.facebook.LoggingBehavior
import com.facebook.appevents.AppEventsConstants
import com.facebook.appevents.AppEventsLogger
import com.facebook.internal.AttributionIdentifiers
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.math.BigDecimal
import java.util.Currency

class HybridFacebookAnalytics : HybridFacebookAnalyticsSpec() {

    private val logger: AppEventsLogger by lazy {
        AppEventsLogger.newLogger(FacebookSdk.getApplicationContext())
    }

    // Background scope for blocking SDK calls (Play Services I/O) so they never
    // run on the JS thread.
    private val ioScope = CoroutineScope(Dispatchers.IO)

    // region Initialization & Runtime Configuration

    override fun initialize() {
        FacebookSdk.setAutoInitEnabled(true)
        FacebookSdk.fullyInitialize()
    }

    override fun setAutoLogAppEventsEnabled(enabled: Boolean) {
        FacebookSdk.setAutoLogAppEventsEnabled(enabled)
    }

    override fun setAdvertiserIDCollectionEnabled(enabled: Boolean) {
        FacebookSdk.setAdvertiserIDCollectionEnabled(enabled)
    }

    override fun setAppID(appID: String) {
        FacebookSdk.setApplicationId(appID)
    }

    override fun setClientToken(clientToken: String) {
        FacebookSdk.setClientToken(clientToken)
    }

    override fun setLoggingEnabled(enabled: Boolean) {
        val behaviors = listOf(
            LoggingBehavior.APP_EVENTS,
            LoggingBehavior.NETWORK_REQUESTS,
            LoggingBehavior.DEVELOPER_ERRORS,
            LoggingBehavior.INCLUDE_ACCESS_TOKENS
        )
        FacebookSdk.setIsDebugEnabled(enabled)
        for (behavior in behaviors) {
            if (enabled) {
                FacebookSdk.addLoggingBehavior(behavior)
            } else {
                FacebookSdk.removeLoggingBehavior(behavior)
            }
        }
    }

    // endregion

    // region Event Logging

    override fun logEvent(eventName: String, valueToSum: Double, params: Map<String, String>) {
        val bundle = mapToBundle(params)
        logger.logEvent(eventName, valueToSum, bundle)
    }

    override fun logEventWithoutParams(eventName: String, valueToSum: Double) {
        if (valueToSum != 0.0) {
            logger.logEvent(eventName, valueToSum)
        } else {
            logger.logEvent(eventName)
        }
    }

    override fun logPurchase(amount: Double, currency: String, params: Map<String, String>) {
        val bundle = mapToBundle(params)
        logger.logPurchase(
            BigDecimal.valueOf(amount),
            Currency.getInstance(currency),
            bundle
        )
    }

    override fun logPushNotificationOpen(payload: Map<String, String>) {
        val bundle = mapToBundle(payload)
        logger.logPushNotificationOpen(bundle)
    }

    // endregion

    // region User Identity

    override fun setUserID(userID: String) {
        AppEventsLogger.setUserID(userID)
    }

    override fun clearUserID() {
        AppEventsLogger.clearUserID()
    }

    override fun getUserID(): String? {
        return AppEventsLogger.getUserID()
    }

    override fun setUserData(
        email: String?,
        firstName: String?,
        lastName: String?,
        phone: String?,
        dateOfBirth: String?,
        gender: String?,
        city: String?,
        state: String?,
        zip: String?,
        country: String?
    ) {
        AppEventsLogger.setUserData(
            email, firstName, lastName, phone,
            dateOfBirth, gender, city, state, zip, country
        )
    }

    // endregion

    // region Device Identifiers

    override fun getAnonymousID(): Promise<String?> {
        return Promise.resolved(logger.anonymousAppDeviceGUID)
    }

    override fun getAdvertiserID(): Promise<String?> {
        // AttributionIdentifiers performs blocking Play Services I/O, so run it
        // off the JS thread on the IO dispatcher.
        return Promise.async(ioScope) {
            try {
                val context = FacebookSdk.getApplicationContext()
                val identifiers = AttributionIdentifiers.getAttributionIdentifiers(context)
                identifiers?.androidAdvertiserId
            } catch (e: Exception) {
                null
            }
        }
    }

    // endregion

    // region Configuration

    override fun setFlushBehavior(flushBehavior: String) {
        when (flushBehavior) {
            "explicit_only" -> AppEventsLogger.setFlushBehavior(AppEventsLogger.FlushBehavior.EXPLICIT_ONLY)
            else -> AppEventsLogger.setFlushBehavior(AppEventsLogger.FlushBehavior.AUTO)
        }
    }

    override fun flush() {
        logger.flush()
    }

    // endregion

    // region Push Notifications

    override fun setPushNotificationsDeviceToken(token: String) {
        // No-op on Android — use setPushNotificationsRegistrationId instead
    }

    override fun setPushNotificationsRegistrationId(registrationId: String) {
        AppEventsLogger.setPushNotificationsRegistrationId(registrationId)
    }

    // endregion

    // region Helpers

    private fun mapToBundle(map: Map<String, String>): Bundle {
        val bundle = Bundle()
        for ((key, value) in map) {
            bundle.putString(key, value)
        }
        return bundle
    }

    // endregion
}
